### PR TITLE
Namespace override

### PR DESCRIPTION
**Pull request by kfox1111**
This patch makes it possible to install the subcharts in different namespaces
as needed.

fixes https://github.com/spiffe/helm-charts/issues/17

### CODE REVIEW

File: .github/workflows/helm-chart-ci.yaml [hunk 1]
```diff
@@ -151,17 +151,19 @@ jobs:
...
         run: |
+          export EXTRA_HELM_ARGS=""
           [ "$VALUES" != "default" ] && kubectl create namespace "$VALUES"
           post-install() {
             [ -x .github/tests/$VALUES/post-install.sh ] && .github/tests/$VALUES/post-install.sh
             exit $1
           }
           trap 'post-install $? $LINENO' EXIT
           [ -x .github/tests/$VALUES/pre-install.sh ] && .github/tests/$VALUES/pre-install.sh
+          [ -x .github/tests/$VALUES/settings.sh ] && source .github/tests/$VALUES/settings.sh
```
Comment (marcofranssen) [inline .github/workflows/helm-chart-ci.yaml:162]:
```suggestion
          echo "${EXTRA_HELM_ARGS}"
          [ -x .github/tests/$VALUES/settings.sh ] && source .github/tests/$VALUES/settings.sh
          echo "${EXTRA_HELM_ARGS}"
```

Maybe add some temporary commit with debugging lines.

You can later drop that commit with interactive rebase

```shell
git rebase -i main
```

Then in editor change `pick` to `d` for that particular commit and close the editor.

**Comment by faisal-memon** [comment]
I see some other stuff around lockdown, securitycontext, etc. mixed in here. Can we focus this pr purely on the namespace issue? Sorry im going to be a stickler on that one.

**Comment by kfox1111** [comment]
yes/no. Its based on some other PR's, so it can be tested. But there is no way not to show those other pr's in this one so it won't look clean until the other pr's merge. Same thing is happening with the lockdown pr. as stuff merges, it looks more and more simple/focused. We will get there. needs a bit of time to review other prs, such as https://github.com/spiffe/helm-charts/pull/93

**Comment by faisal-memon** [comment]
> yes/no. Its based on some other PR's, so it can be tested. But there is no way not to show those other pr's in this one so it won't look clean until the other pr's merge. Same thing is happening with the lockdown pr. as stuff merges, it looks more and more simple/focused. We will get there. needs a bit of time to review other prs, such as #93
 
I think this one is very simple to split off Im happy to take it over if you have too much on your plate.

**Comment by kfox1111** [comment]
Its really easy to commit broken. I caught half a dozen issues with the pr once I started gate testing it. We need to merge some of the other bits first.

**Comment by kfox1111** [comment]
I think we're talking past each other. It will be broken out and merged as a more focused pr. but its part of a chain of prs that ensure it actually works. That's why its still Draft.

File: .github/tests/lockdown/post-install.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+if [ $1 -ne 0 ]; then
+  echo '==> Events of namespace spire-server'
```
Comment (marcofranssen) [inline .github/tests/lockdown/post-install.sh:3]: Better to write this to GITHUB_STEP_SUMMARY like I do in my draft pr. You can format it in markdown code blocks.

File: charts/spire/charts/spiffe-oidc-discovery-provider/templates/tests/test-connection.yaml [hunk 3]
```diff
@@ -19,13 +20,13 @@ spec:
     - name: wget-service-name-namespace
       image: busybox
       command: ['wget']
-      args: ['-O', '/dev/null', '{{ include "spiffe-oidc-discovery-provider.fullname" . }}.{{ .Release.Namespace }}:{{ .Values.service.port }}/.well-known/openid-configuration']
+      args: ['-O', '/dev/null', '{{ include "spiffe-oidc-discovery-provider.fullname" . }}.{{ include "spiffe-oidc-discovery-provider.namespace" . }}:{{ .Values.service.port }}/.well-known/openid-configuration']
```
Comment (marcofranssen) [inline charts/spire/charts/spiffe-oidc-discovery-provider/templates/tests/test-connection.yaml:23]: Why hiding test output? Makes it harder to debug.

Comment (kfox1111) [reply charts/spire/charts/spiffe-oidc-discovery-provider/templates/tests/test-connection.yaml:23]: That part is from the other review. The only thing this does is change the namespace variable.

File: .github/tests/lockdown/post-install.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+k_rollout_status=(kubectl rollout status --watch --timeout 30s --namespace)
+
+cat <<EOF >>"$GITHUB_STEP_SUMMARY"
+### spire
+| workload | Status |
+| -------- | ------ |
+| spire-server | $("${k_rollout_status[@]}" spire-server statefulset spire-server) |
+| spire-spiffe-csi-driver | $("${k_rollout_status[@]}" spire-system daemonset spire-spiffe-csi-driver) |
+| spire-agent | $("${k_rollout_status[@]}" spire-server daemonset spire-agent) |
+| spire-spiffe-oidc-discovery-provider | $("${k_wait[@]}" spire-server deployments.apps spire-spiffe-oidc-discovery-provider) |
```
Comment (marcofranssen) [inline .github/tests/lockdown/post-install.sh:18]: I liked your events post as well. Maybe include that in a codeblock so it renders nicely. Three backtracks with shell highlighting.

Also check the namespace. Is that matching with what you need?

File: charts/spire/charts/spire-agent/templates/_helpers.tpl [hunk 5]
```diff
@@ -23,6 +23,25 @@ If release name contains chart name it will be used as a full name.
...
+{{- end -}}
+
+{{- define "spire-agent.server.namespace" -}}
+  {{- if .Values.server.namespaceOverride -}}
+    {{- .Values.server.namespaceOverride -}}
+  {{- else -}}
+    {{- .Release.Namespace -}}
+  {{- end -}}
+{{- end -}}
+
```
Comment (faisal-memon) [inline charts/spire/charts/spire-agent/templates/_helpers.tpl:44]: This seems duplicated to me? Am I missing something?

Comment (faisal-memon) [reply charts/spire/charts/spire-agent/templates/_helpers.tpl:44]: Nm, the second block is for the server namespace.

**Comment by faisal-memon** [comment]
lgtm overall.

File: .github/tests/lockdown/post-install.sh (hunk 4 above)
Comment (kfox1111) [reply .github/tests/lockdown/post-install.sh:18]: Done

**Review by faisal-memon — APPROVED** [review]

**Comment by kfox1111** [comment]
Fixed the merge conflict.

**Pull request merged**